src/partition/repart.c

```diff
@@ -2982,23 +3146,39 @@ static int partition_encrypt(
         if (r < 0)
                 return log_error_errno(r, "Failed to generate volume key: %m");
 
-        r = sym_crypt_init(&cd, node);
+        r = fopen_temporary(NULL, &h, &hp);
         if (r < 0)
-                return log_error_errno(r, "Failed to allocate libcryptsetup context: %m");
+                return log_error_errno(r, "Failed to create temporary LUKS header file: %m");
+
+        /* Weird cryptsetup requirement which requires the header file to be the size of at least one sector. */
+        r = posix_fallocate(fileno(h), 0, context->sector_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to grow temporary LUKS header file: %m");
+
+        r = sym_crypt_init(&cd, hp);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate libcryptsetup context for %s: %m", hp);
 
         cryptsetup_enable_logging(cd);
 
+        r = sym_crypt_metadata_locking(cd, false);
+        if (r < 0)
+                return log_error_errno(r, "Failed to disable metadata locking: %m");
+
+        /* I don't really get why we have to divide by 2 here but it's the same as what cryptsetup does to
+         * make offline encryption work. */
+        r = sym_crypt_set_data_offset(cd, LUKS2_METADATA_SIZE / (2 * context->sector_size));
```

Comment:
So what does this need to be then? I copied this approach from the cryptsetup sources

Reply:
data_offset value may be same as (absolute) data_shift value passed in reencryption parameters. But there has to be twice the (absolute) data_shift value of spare data area at the end of the image file.

This is what encryption initialization (when used from cryptsetup) does to the data device. It moves original first segment to the end of data device and puts LUKS2 header in head of the device:

```
[LUKS2_header] [data_segment 2].....[data_segment 'n'][data_shift size spare area][data_segment 1]
```

Reply:
So is the current version where we set `data_shift = LUKS2_METADATA_SIZE / (2 * 512)` and `data_offset = LUKS2_METADATA_SIZE / (2 * 512)` correct if we keep `LUKS2_METADATA_SIZE` free space at the end of the partition? `LUKS2_METADATA_SIZE` is 16M. It seems to be the same as what `cryptsetup` does.

Reply:
You may have detached LUKS header. Aka header stored on different device/file from actual data device/file.  With detached header you don't need additional free space at all.

Only offline encryption with header put on data device requires free space. All other operations may be performed without extra free space on data device.

Reply:
But why do we need free space 2x the luks header size? I get why we need 1x data shift free space so we can shift the data to make space for the header. But what about the rest of the free space? Isn't that just scratch space for reencryption. Couldn't that scratch space be allocated in another file?

Reply:
This how the device looks like before the encryption:
```
[data_segment 1 (size = X)][data_segment 2]...[data_segment 'n'][ free space (size = 2X)]
```
This is how the data looks like after device encryption gets initialized:
```
[LUKS2_header (size = X)] [data_segment 2].....[data_segment 'n'][ free space (size = X) ][data_segment 1 (size = X)]
```
During the encryption, data gets shifted backwards so that, in the end, data_segment 1 can be put right after the LUKS2 header as with any other LUKS device.

The result looks like follows:
```
[LUKS2_header (size = X)] [data_segment 1 (size = X)][data_segment 2]....[data_segment 'n'][ free space (size = X) ]
```

Reply:
> Couldn't that scratch space be allocated in another file?

Could be. But nobody implemented that feature yet :)

Reply:
Cool, we'll have to use a bit more space for now then. Thanks for answering all my questions! We should be doing offline encryption correctly now.